File: calicam_mono.cpp

// This is NWNC (No Warranty No Copyright) Software.
// astar.ai
// Nov 17, 2019

#include <opencv2/opencv.hpp>

#define MARGIN    15. * CV_PI / 180.

////////////////////////////////////////////////////////////////////////////////

bool      live = false;
//bool      live = true;
//To run live mode, you need a CaliCam from www.astar.ai

int       vfov_bar = 60, size_bar =   0;
int       vfov_max = 60, size_max = 480;
int       vfov_now = 120, size_now = 480;

int       cap_cols, cap_rows;
bool      changed = false;
cv::Mat   Kl, Dl, xil, fmap[2];

////////////////////////////////////////////////////////////////////////////////

enum RectMode {
  RECT_PERSPECTIVE,
  RECT_CYLINDRICAL,
  RECT_FISHEYE,
  RECT_LONGLAT
};

RectMode mode = RECT_PERSPECTIVE;

////////////////////////////////////////////////////////////////////////////////

void OnTrackAngle(int value, void*) {
  vfov_bar = value;
  vfov_now = 60 + vfov_bar;
  changed = true;
}

////////////////////////////////////////////////////////////////////////////////

void OnTrackSize(int value, void*) {
  size_bar = value;
  size_now = 480 + size_bar;
  if (size_now % 2 == 1)
    size_now--;
  changed = true;
}

////////////////////////////////////////////////////////////////////////////////

void LoadParameters(std::string file_name) {
  cv::FileStorage fs(file_name, cv::FileStorage::READ);
  if (!fs.isOpened()) {
    std::cout << "Failed to open ini parameters" << std::endl;
    exit(-1);
  }

  cv::Size cap_size;
  fs["cap_size" ] >> cap_size;
  fs["Kl"       ] >> Kl;
  fs["Dl"       ] >> Dl;
  fs["xil"      ] >> xil;

  fs.release();

  cap_cols  = cap_size.width;
  cap_rows  = cap_size.height;
}

////////////////////////////////////////////////////////////////////////////////

inline double MatRowMul(cv::Mat m, double x, double y, double z, int r) {
  return m.at<double>(r,0) * x + m.at<double>(r,1) * y + m.at<double>(r,2) * z;
}

////////////////////////////////////////////////////////////////////////////////

void InitRectifyMap(cv::Mat K,
                    cv::Mat D,
                    cv::Mat Knew,
                    double xi,
                    cv::Size size,
                    RectMode mode,
                    cv::Mat& map1,
                    cv::Mat& map2) {
  map1.create(size, CV_32F);
  map2.create(size, CV_32F);

  double fx = K.at<double>(0,0);
  double fy = K.at<double>(1,1);
  double cx = K.at<double>(0,2);
  double cy = K.at<double>(1,2);
  double s  = K.at<double>(0,1);

  double k1 = D.at<double>(0,0);
  double k2 = D.at<double>(0,1);
  double p1 = D.at<double>(0,2);
  double p2 = D.at<double>(0,3);

  cv::Mat Ki  = Knew.inv();

  for (int r = 0; r < size.height; ++r) {
    for (int c = 0; c < size.width; ++c) {
      double xc = 0.;
      double yc = 0.;
      double zc = 0.;

      if (mode == RECT_PERSPECTIVE) {
        xc = MatRowMul(Ki, c, r, 1., 0);
        yc = MatRowMul(Ki, c, r, 1., 1);
        zc = MatRowMul(Ki, c, r, 1., 2);
      }

      if (mode == RECT_CYLINDRICAL) {
        double tt = MatRowMul(Ki, c, r, 1., 0);
        double pp = MatRowMul(Ki, c, r, 1., 1) + MARGIN;

        xc = -sin(pp) * cos(tt);
        yc = -cos(pp);
        zc =  sin(pp) * sin(tt);
      }

      if (mode == RECT_FISHEYE) {
        if (hypot(c - Knew.at<double>(0, 2), r - Knew.at<double>(1, 2))
            > Knew.at<double>(0, 0)) {
          map1.at<float>(r,c) = -1.f;
          map2.at<float>(r,c) = -1.f;
          continue;
        }
        double ee = MatRowMul(Ki, c, r, 1., 0);
        double ff = MatRowMul(Ki, c, r, 1., 1);

        double zz = 2. / (ee * ee + ff * ff + 1.);

        xc = zz * ee;
        yc = zz * ff;
        zc = zz - 1.;
      }

      if (mode == RECT_LONGLAT) {
        double tt = MatRowMul(Ki, c, r, 1., 0);
        double pp = MatRowMul(Ki, c, r, 1., 1);

        xc = -cos(tt);
        yc = -sin(tt) * cos(pp);
        zc =  sin(tt) * sin(pp);
      }

      double rr = sqrt(xc * xc + yc * yc + zc * zc);
      double xs = xc / rr;
      double ys = yc / rr;
      double zs = zc / rr;

      double xu = xs / (zs + xi);
      double yu = ys / (zs + xi);

      double r2 = xu * xu + yu * yu;
      double r4 = r2 * r2;
      double xd = (1+k1*r2+k2*r4)*xu + 2*p1*xu*yu + p2*(r2+2*xu*xu);
      double yd = (1+k1*r2+k2*r4)*yu + 2*p2*xu*yu + p1*(r2+2*yu*yu);

      double u = fx * xd + s * yd + cx;
      double v = fy * yd + cy;

      map1.at<float>(r,c) = (float) u;
      map2.at<float>(r,c) = (float) v;
    }
  }
}

////////////////////////////////////////////////////////////////////////////////

void InitRectifyMap() {
  double   vfov_rad, focal;
  cv::Mat  Knew;
  cv::Size img_size(size_now, size_now);

  switch (mode) {
  case RECT_PERSPECTIVE:
    std::cout << "\x1b[1;36m" << "Mode: " << "Perspective" << "\x1b[0m\n";
    vfov_rad = vfov_now * CV_PI / 180.;
    focal = size_now / 2. / tan(vfov_rad / 2.);
    Knew = (cv::Mat_<double>(3, 3) << focal, 0., size_now  / 2. - 0.5,
                                      0., focal, size_now / 2. - 0.5,
                                      0., 0., 1.);
    InitRectifyMap(Kl, Dl, Knew, xil.at<double>(0,0),
                   img_size, RECT_PERSPECTIVE, fmap[0], fmap[1]);

    std::cout << "V.Fov: "  << vfov_now   << "\t";
    break;

  case RECT_CYLINDRICAL:
    std::cout << "\x1b[1;36m" << "Mode: " << "Cylindrical" << "\x1b[0m\n";
    Knew = cv::Mat::eye(3, 3, CV_64F);
    Knew.at<double>(0,0) = size_now  /  CV_PI;
    Knew.at<double>(1,1) = size_now / (CV_PI - 2 * MARGIN);
    InitRectifyMap(Kl, Dl, Knew, xil.at<double>(0,0),
                   img_size, RECT_CYLINDRICAL, fmap[0], fmap[1]);
    break;

  case RECT_FISHEYE:
    img_size = cv::Size(size_now, size_now);

    std::cout << "\x1b[1;36m" << "Mode: " << "Fisheye" << "\x1b[0m\n";
    Knew = cv::Mat::eye(3, 3, CV_64F);
    Knew.at<double>(0,0) = size_now  / 2.;
    Knew.at<double>(0,2) = size_now  / 2. - 0.5;
    Knew.at<double>(1,1) = size_now  / 2.;
    Knew.at<double>(1,2) = size_now / 2. - 0.5;
    InitRectifyMap(Kl, Dl, Knew, xil.at<double>(0,0),
                   img_size, RECT_FISHEYE, fmap[0], fmap[1]);
    break;

  case RECT_LONGLAT:
    std::cout << "\x1b[1;36m" << "Mode: " << "Longitude-Latitude" << "\x1b[0m\n";
    Knew = cv::Mat::eye(3, 3, CV_64F);
    Knew.at<double>(0,0) = size_now  / CV_PI;
    Knew.at<double>(1,1) = size_now / CV_PI;
    InitRectifyMap(Kl, Dl, Knew, xil.at<double>(0,0),
                   img_size, RECT_LONGLAT, fmap[0], fmap[1]);
    break;
  }

  std::cout << "Width: "  << size_now  << "\t"
            << "Height: " << size_now << std::endl;
  std::cout << "K Matrix: \n" << Knew << std::endl << std::endl;
}

////////////////////////////////////////////////////////////////////////////////

int main(int argc, char** argv) {
  std::string param_name = "../astar_calicam_mono.yml";
  std::string image_name = "../times_square.jpg";

  if (argc == 2) {
    param_name = argv[1];
  } else if (argc == 3) {
    param_name = argv[1];
    image_name = argv[2];
  }

  LoadParameters(param_name);
  InitRectifyMap();

  cv::Mat raw_img;
  cv::VideoCapture vcapture;
  if (live) {
    vcapture.open(0);

    if (!vcapture.isOpened()) {
      std::cout << "Camera doesn't work" << std::endl;
      exit(-1);
    }

    vcapture.set(cv::CAP_PROP_FRAME_WIDTH,  cap_cols);
    vcapture.set(cv::CAP_PROP_FRAME_HEIGHT, cap_rows);
    vcapture.set(cv::CAP_PROP_FPS, 30);
  } else {
    raw_img = cv::imread(image_name, cv::IMREAD_COLOR);
  }

  char win_name[256];
  sprintf(win_name, "Raw Image: %d x %d", cap_cols, cap_rows);
  std::string param_win_name(win_name);
  cv::namedWindow(param_win_name);

  cv::createTrackbar("V. FoV:  60    +", param_win_name, nullptr,   vfov_max,   OnTrackAngle);
  cv::setTrackbarPos("V. FoV:  60    +", param_win_name, vfov_bar);
  cv::createTrackbar("Size:  480 +", param_win_name, nullptr,  size_max,  OnTrackSize);

  cv::Mat raw_imgl, raw_imgr, rect_imgl, rect_imgr;
  while (1) {
    if (changed) {
      InitRectifyMap();
      changed = false;
    }

    if (live)
      vcapture >> raw_img;

    if (raw_img.total() == 0) {
      std::cout << "Image capture error" << std::endl;
      exit(-1);
    }

    raw_imgl = raw_img;
    cv::remap(raw_img, rect_imgl, fmap[0], fmap[1], 1, 0);

    cv::Mat small_img;
    cv::resize(raw_imgl, small_img, cv::Size(), 0.5, 0.5);
    imshow(param_win_name, small_img);
    imshow("Rectified Image", rect_imgl);

    char key = cv::waitKey(1);

    if (key == '1') {
      mode = RECT_PERSPECTIVE;
      changed = true;
    }

    if (key == '2') {
      mode = RECT_CYLINDRICAL;
      changed = true;
    }

    if (key == '3') {
      mode = RECT_FISHEYE;
      changed = true;
    }

    if (key == '4') {
      mode = RECT_LONGLAT;
      changed = true;
    }

    if (key == 32) {
      imwrite("img.jpg", rect_imgl);
    }

    if (key == 'q' || key == 'Q' || key == 27)
      break;
  }

  return 0;
}

////////////////////////////////////////////////////////////////////////////////

File: calicam_mono.py

import sys
import cv2
import math
import numpy as np

#-------------------------------------------------------------------------------#

Kl = None
Dl = None
xil = None

mapx = None
mapy = None

cap_cols = None
cap_rows = None

vfov_bar =  0
width_bar = 0
height_bar = 0
vfov_max = 60
width_max = 480
height_max = 360
vfov_now = 60
width_now = 480
height_now = 360

changed = False

margin = 15. * math.pi / 180.

mode = 'kRectPerspective'

#-------------------------------------------------------------------------------#

def load_parameters(param_file):
  global Kl, Dl, xil
  global cap_cols, cap_rows
  
  fs = cv2.FileStorage(param_file, cv2.FILE_STORAGE_READ)
  
  cap_size_node = fs.getNode("cap_size")
  cap_cols = int(cap_size_node.at(0).real())
  cap_rows = int(cap_size_node.at(1).real())
  
  Kl = fs.getNode("Kl").mat()
  Dl = fs.getNode("Dl").mat()
  xil = fs.getNode("xil").mat()
  
#-------------------------------------------------------------------------------#

def init_undistort_rectify_map(k, d, r, knew, xi0, size, mode):
  fx = k[0, 0]
  fy = k[1, 1]
  cx = k[0, 2]
  cy = k[1, 2]
  s  = k[0, 1]
  
  k1 = d[0, 0]
  k2 = d[0, 1]
  p1 = d[0, 2]
  p2 = d[0, 3]
  
  ki = np.linalg.inv(knew)
  ri = np.linalg.inv(r)
  kri = np.linalg.inv(np.matmul(knew, r))
  
  rows = size[0]
  cols = size[1]
  
  mapx = np.zeros((rows, cols), dtype = np.float32)
  mapy = np.zeros((rows, cols), dtype = np.float32)
  
  print("Wait, this takes a while ... ")
  for r in range(rows):
    for c in range(cols):
      xc = 0.0
      yc = 0.0
      zc = 0.0
      
      if mode == 'kRectPerspective':
        cr1 = np.array([c, r, 1.])
        xc = np.dot(kri[0, :], cr1)
        yc = np.dot(kri[1, :], cr1)
        zc = np.dot(kri[2, :], cr1)
  
      if mode == 'kRectLonglat':
        tt = (c * 1. / (cols - 1) - 0.5) * math.pi
        pp = (r * 1. / (rows - 1) - 0.5) * math.pi

        xn = math.sin(tt)
        yn = math.cos(tt) * math.sin(pp)
        zn = math.cos(tt) * math.cos(pp)
      
        cr1 = np.array([xn, yn, zn])
        xc = np.dot(ri[0, :], cr1)
        yc = np.dot(ri[1, :], cr1)
        zc = np.dot(ri[2, :], cr1)
  
      if mode == 'kRectFisheye':
        cr1 = np.array([c, r, 1.])
        ee = np.dot(ki[0, :], cr1)
        ff = np.dot(ki[1, :], cr1)
        zz = 2. / (ee * ee + ff * ff + 1.)

        xn = zz * ee
        yn = zz * ff
        zn = zz - 1.

        cr1 = np.array([xn, yn, zn])
        xc = np.dot(ri[0, :], cr1)
        yc = np.dot(ri[1, :], cr1)
        zc = np.dot(ri[2, :], cr1)
        
      if mode == 'kRectCylindrical':
        cr1 = np.array([c, r, 1.])
        tt = np.dot(ki[0, :], cr1)
        pp = np.dot(ki[1, :], cr1) + margin

        xc = -math.sin(pp) * math.cos(tt)
        yc = -math.cos(pp)
        zc =  math.sin(pp) * math.sin(tt)
        
      if zc < 0.0:
        mapx[r, c] = np.float32(-1.)
        mapy[r, c] = np.float32(-1.)

        continue
      
      rr = math.sqrt(xc * xc + yc * yc + zc * zc)
      xs = xc / rr
      ys = yc / rr
      zs = zc / rr
      
      xu = xs / (zs + xi0)
      yu = ys / (zs + xi0)
      
      r2 = xu * xu + yu * yu
      r4 = r2 * r2
      xd = (1 + k1 * r2 + k2 * r4) * xu + 2 * p1 * xu * yu + p2 * (r2 + 2 * xu * xu)
      yd = (1 + k1 * r2 + k2 * r4) * yu + 2 * p2 * xu * yu + p1 * (r2 + 2 * yu * yu)

      u = fx * xd + s * yd + cx
      v = fy * yd + cy

      mapx[r, c] = np.float32(u)
      mapy[r, c] = np.float32(v)

  return mapx, mapy

#-------------------------------------------------------------------------------#

def init_rectify_map():
  global mode, mapx, mapy

  Rl = np.identity(3, dtype = np.float64)
  Knew = None
  
  if mode == 'kRectPerspective':
    print('kRectPerspective')

    vfov_rad = vfov_now * math.pi / 180.
    focal = height_now / 2. / math.tan(vfov_rad / 2.)
    
    Knew = np.identity(3, dtype = np.float64)
    Knew[0, 0] = focal
    Knew[1, 1] = focal
    Knew[0, 2] = width_now / 2 - 0.5
    Knew[1, 2] = height_now / 2 - 0.5

    img_size = [height_now, width_now]
    mapx, mapy = init_undistort_rectify_map(Kl, Dl, Rl, Knew, xil, img_size, 'kRectPerspective')
  
    print('Width: {}, Height: {}, V.FoV: {}'.format(width_now, height_now, vfov_now))
    
  if mode == 'kRectLonglat':
    print('kRectLonglat')

    img_length = width_now if width_now > height_now else height_now
    Knew = np.identity(3, dtype = np.float64)
    Knew[0, 0] = img_length / math.pi
    Knew[1, 1] = img_length / math.pi
    
    img_size = [img_length, img_length]
    mapx, mapy = init_undistort_rectify_map(Kl, Dl, Rl, Knew, xil, img_size, 'kRectLonglat')

    print('Width: {}, Height: {}'.format(img_length, img_length))
    
  if mode == 'kRectFisheye':
    print('kRectFisheye')

    img_length = width_now if width_now > height_now else height_now
    Knew = np.identity(3, dtype = np.float64)
    Knew[0, 0] = img_length / 2
    Knew[1, 1] = img_length / 2
    Knew[0, 2] = img_length / 2 - 0.5
    Knew[1, 2] = img_length / 2 - 0.5

    img_size = [img_length, img_length]
    mapx, mapy = init_undistort_rectify_map(Kl, Dl, Rl, Knew, xil, img_size, 'kRectFisheye')

    print('Width: {}, Height: {}'.format(img_length, img_length))
    
  if mode == 'kRectCylindrical':
    print('kRectCylindrical')

    img_length = width_now if width_now > height_now else height_now
    Knew = np.identity(3, dtype = np.float64)

    Knew[0, 0] = img_length / math.pi
    Knew[1, 1] = img_length / (math.pi - 2 * margin)
    
    img_size = [img_length, img_length]
    mapx, mapy = init_undistort_rectify_map(Kl, Dl, Rl, Knew, xil, img_size, 'kRectCylindrical')

    print('Width: {}, Height: {}'.format(img_length, img_length))
    
  print('K Matrix:')
  print(Knew)
  print('')
  
#-------------------------------------------------------------------------------#

def OnTrackAngle(vfov_bar):
  global vfov_now, changed
  vfov_now = 60 + vfov_bar
  changed = True

#-------------------------------------------------------------------------------#

def OnTrackWidth(width_bar):
  global width_now, changed
  width_now = 480 + width_bar
  if width_now % 2 == 1:
    width_now = width_now - 1
  changed = True

#-------------------------------------------------------------------------------#

def OnTrackHeight(height_bar):
  global height_now, changed
  height_now = 360 + height_bar
  if height_now % 2 == 1:
    height_now = height_now - 1
  changed = True

#-------------------------------------------------------------------------------#

def main():  
  global changed, mode, mapx, mapy
  
  param_file = "astar_calicam_mono.yml"
  image_name = "dasl_wood_shop_mono.jpg"

  if len(sys.argv) == 2:
    param_file = sys.argv[1]

  if len(sys.argv) == 3:
    param_file = sys.argv[1]
    image_name = sys.argv[2]
  
  load_parameters(param_file)
  init_rectify_map()
  
  raw_img = cv2.imread(image_name, 1)
  
  param_win_name = "Raw Image: " + str(cap_cols) + " x " + str(cap_rows)
  
  cv2.namedWindow(param_win_name)
  cv2.createTrackbar("V. FoV:  60    +", param_win_name, vfov_bar, vfov_max, OnTrackAngle)
  cv2.createTrackbar("Width:  480 +", param_win_name, width_bar, width_max, OnTrackWidth)
  cv2.createTrackbar("Height: 360 +", param_win_name, height_bar, height_max, OnTrackHeight)
 
  while True:
    if changed == True:
      init_rectify_map()
      changed = False
    
    raw_imgl = raw_img
    rect_imgl = cv2.remap(raw_imgl, mapx, mapy, cv2.INTER_LINEAR)
    
    dim = (cap_cols / 2, cap_rows / 2)
    small_img = cv2.resize(raw_imgl, dim, interpolation = cv2.INTER_NEAREST)
    
    cv2.imshow(param_win_name, small_img)
    cv2.imshow("Rectified Image", rect_imgl)
    
    key = cv2.waitKey(1)
    
    if key & 0xFF == ord('1'):
      mode = 'kRectPerspective'
      changed = True
    
    if key & 0xFF == ord('2'):
      mode = 'kRectCylindrical'
      changed = True
    
    if key & 0xFF == ord('3'):
      mode = 'kRectFisheye'
      changed = True
    
    if key & 0xFF == ord('4'):
      mode = 'kRectLonglat'
      changed = True
    
    if key & 0xFF == ord('q') or key  == 27:
      break

#-------------------------------------------------------------------------------#

if __name__ == "__main__":
  main()

#-------------------------------------------------------------------------------#

